src/addons/MountNode/MountNode.js

```diff
@@ -4,36 +4,37 @@ import { Component } from 'react'
 import { customPropTypes } from '../../lib'
 import getNodeRefFromProps from './lib/getNodeRefFromProps'
 import handleClassNamesChange from './lib/handleClassNamesChange'
-import NodeRegistry from './lib/NodeRegistry'
-
-const nodeRegistry = new NodeRegistry()
 
 /**
  * A component that allows to manage classNames on a DOM node in declarative manner.
  */
 export default class MountNode extends Component {
+  constructor(props) {
+    super(props)
+
+    this.ref = getNodeRefFromProps(props)
+  }
+
   shouldComponentUpdate({ className: nextClassName }) {
     const { className: currentClassName } = this.props
 
     return nextClassName !== currentClassName
   }
 
   componentDidMount() {
-    const nodeRef = getNodeRefFromProps(this.props)
-
-    nodeRegistry.add(nodeRef, this)
-    nodeRegistry.emit(nodeRef, handleClassNamesChange)
+    handleClassNamesChange(this.ref, [this])
   }
 
-  componentDidUpdate() {
-    nodeRegistry.emit(getNodeRefFromProps(this.props), handleClassNamesChange)
+  componentDidUpdate(prevProps) {
+    if (this.props.node !== prevProps.node) {
+      this.ref = getNodeRefFromProps(this.props)
+    }
+
+    handleClassNamesChange(this.ref, [this])
   }
 
   componentWillUnmount() {
-    const nodeRef = getNodeRefFromProps(this.props)
-
-    nodeRegistry.del(nodeRef, this)
-    nodeRegistry.emit(nodeRef, handleClassNamesChange)
```

Comment:
I don't think that it's valid to remove this part as it solves duplicated classnames issue, for example:

```tsx
<MountNode className="foo" node={document.body} />
<MountNode className="foo" node={document.body} />
<MountNode className="bar" node={document.body} />
```

Should add only `foo bar`.

This also handles unmount scenario: when one of `MountNode` components with `className="foo"` will be removed - `foo` should still be defined on `document.body` as there is another component that defines it.

Reply:
Got it; thanks for this additional test case. I will take a look and see if my change can be reworked.

Reply:
@jluxenberg do not spend cycles on `MountNode`, I am working on a hook that will replace it...

Reply:
@jluxenberg FYI: #4027

Reply:
I merged the original PR and solved merge conflicts in this. Can you please check? It also seems that we have a single comment left 👍